lib/chef/cookbook_loader.rb

```diff
@@ -66,17 +66,27 @@ def merged_cookbook_paths # for deprecation warnings
       merged_cookbook_paths
     end
 
+    def warn_about_cookbook_shadowing
+      unless merged_cookbooks.empty?
+        Chef::Log.deprecation "The cookbook(s): #{merged_cookbooks.join(', ')} exist in multiple places in your cookbook_path. " +
+          "A composite version has been compiled.  In a future version of Chef this behavior will be removed."
```

Comment:
We could throw in an option to turn it off, in `#initialize` or something. Generally I'm the biggest hater of methods like `def do_something(enable_disable_some_weird_behavior = true)`, but it's easy enough to nuke when 13.0 time comes. Or else you can have `#load_cookbooks_without_shadow_warning` and then have 

```ruby
def load_cookbooks
  result = load_cookbooks_without_shadow_warning
  warn_about_shadowing
  result
end
```